_posts/2019-03-30-numba-stencil.md

```diff
@@ -0,0 +1,323 @@
+just works with systems like Dask, without you having to build in explicit
+support for parallel computing.  This makes the lives of users much easier.
+
+
+Finished result
+---------------
+
+Lets see the full example one more time.
+
+```python
+import numba
+import dask.array as da
+
+@numba.stencil
+def _smooth(x):
+    return (x[-1, -1] + x[-1, 0] + x[-1, 1] +
+            x[ 0, -1] + x[ 0, 0] + x[ 0, 1] +
+            x[ 1, -1] + x[ 1, 0] + x[ 1, 1]) // 9
+
+
+@numba.guvectorize(
+    [(numba.int8[:, :], numba.int8[:, :])],
+    '(n, m) -> (n, m)'
+)
+def smooth(x, out):
+    out[:] = _smooth(x)
+
+x = da.ones((1000000, 1000, 1000), chunks=('auto', -1, -1), dtype='int8')
+smooth(x)
+```
+
+This code is decently approachable by novice users.  They may not understand
+the internal details of gufuncs or Dask arrays or JIT compilation, but they can
+probably copy-paste-and-modify the example above to suit their needs.
+
+The parts that they do want to change are easy to change, like the stencil
+computation, and creating an array of their own data.
+
+This workflow is efficient and scalable, using low-level compiled code and
+potentially clusters of thousands of computers.
+
+
+What could be better
+--------------------
+
+There are a few things that could make this workflow nicer.
+
+1.  It would be nice not to have to specify dtypes in `guvectorize`, but
+    instead specialize to types as they arrive.
+    [numba/numba #2979](https://github.com/numba/numba/issues/2979)
+
+2.  Support GPU accelerators for the stencil computations using
+    [numba.cuda.jit](https://numba.pydata.org/numba-doc/dev/cuda/index.html).
+    Stencil computations are obvious candidates for GPU acceleration, and this
+    is a good accessible point where novice users can specify what they want in
+    a way that is sufficiently constrained for automated systems to rewrite it
+    as CUDA somewhat easily.
+    [numba/numba 3915](https://github.com/numba/numba/issues/3915)
+
+3.  It would have been nicer to be able to apply the `@guvectorize` decorator
+    directly on top of the stencil function like this.
+
+    ```python
+    @numba.guvectorize(...)
+    @numba.stencil
+    def average(x):
+        ...
+    ```
+
+    Rather than have two functions.
+    [numba/numba #3914](https://github.com/numba/numba/issues/3914)
+
+4.  You may have noticed that our guvectorize function had to assign its result into an
+    out parameter.
+
+    ```python
+    @numba.guvectorize(
+        [(numba.int8[:, :], numba.int8[:, :])],
+        '(n, m) -> (n, m)'
+    )
+    def smooth(x, out):
+        out[:] = _smooth(x)
+    ```
+
+    It would have been nicer, perhaps, to just return the output
+
+    ```python
+    def smooth(x):
+        return _smooth(x)
+    ```
+
+    [numba/numba #3916](https://github.com/numba/numba/issues/3916)
+
+5.  The dask-image library could use a `imsave` function
+
+    [dask/dask-image #110](https://github.com/dask/dask-image/issues/110)
```

Comment:
Does `store` or `to_zarr` not work? This is my sense of what people do today.

Reply:
Would you like for me to remove this comment?

Reply:
We can leave it if you prefer. More of a question really.